Create drag start and drag end events for Grid to have access to dragged items on server side

wbadam opened the pull request:
Resolves #8931

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/framework/8950)
<!-- Reviewable:end -->

wbadam committed:
Create grid drag source event to be able to access dragged items on server side (#8931)

wbadam committed:
Create grid drag end event to have access to dragged items on server side (#8931)

wbadam committed:
Change addListener method name so that no castig would be needed

wbadam committed:
Fix drop listener in dnd UI test

pleku commented:
Reviewed 14 of 14 files at r1.
Review status: all files reviewed at latest revision, 20 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIsoOFmE7J9IajLQ3v:-KgIsoOFmE7J9IajLQ3w:b-e8v2z3) (no related file):*
Please remove all mentions of "Server side something", those are IMO totally unnecessary, you should not have to make to distinctions unless you're explaining some difference in between server and client side.

---

*[client/src/main/java/com/vaadin/client/connectors/grid/GridDragSourceExtensionConnector.java, line 102 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIkb0cSK8gN86qKke9:-KgIkb0cSK8gN86qKkeA:bgtnppf) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/client/src/main/java/com/vaadin/client/connectors/grid/GridDragSourceExtensionConnector.java#L102)):*
> ```Java
>         // Collect dragged rows and add their keys to the list
>         getDraggedRows(dragEndEvent).forEach(
>                 row -> draggedItemKeys.add(row.get(GridState.JSONKEY_ROWKEY)));
> ```

Is this really necessary, or should the keys be cached after the drag has started?

Just thinking that what if the grid has been scrolled while dragging, thus the keys could have been removed from client side grid cache, at least in theory.

---

*[client/src/main/java/com/vaadin/client/connectors/grid/GridDragSourceExtensionConnector.java, line 109 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIkUiFOZbV870nRmfL:-KgIkUiFOZbV870nRmfM:bnw7m0t) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/client/src/main/java/com/vaadin/client/connectors/grid/GridDragSourceExtensionConnector.java#L109)):*
> dragStartEvent

It seems to me that this is not always a drag start event now since it is used for drag ends too

---

*[client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java, line 140 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgImfTLI_YdB3lHCJam:-KgImfTLI_YdB3lHCJan:bvuao80) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java#L140)):*
> ```Java
>         // Set text data parameter
>         nativeEvent.getDataTransfer().setData(DragSourceState.DATA_TYPE_TEXT,
>                 getState().dataTransferText);
> ```

Can you isolate this to its own method that can be overridden by extending classes ?

Then you can move from `GridDragSourceExtensionConnector` this bit from `sendDragStartEventToServer`

```java 
        // Collect dragged rows and add their keys to the list
        getDraggedRows(dragStartEvent).forEach(
                row -> draggedItemKeys.add(row.get(GridState.JSONKEY_ROWKEY)));
```

---

*[client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java, line 151 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgImDQz5hOFCglxpuhz:-KgImDQz5hOFCglxpui-:buc4p35) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java#L151)):*
> ```Java
>     /**
>      * Initiates a server RPC for the drag start event.
>      *
> ```

You should probably document that this method is called ONLY IF there is a server side drag start event handler used

---

*[client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java, line 181 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIn2jpPlEYiqMYOTEQ:-KgIn2jqD_6BHIpKKty9:b-efnor9) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java#L181)):*
> start

copy paste error

---

*[server/src/main/java/com/vaadin/event/dnd/DragSourceExtension.java, line 67 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIogQX5BBDGONSZonq:-KgIogQX5BBDGONSZonr:b2hia9o) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/event/dnd/DragSourceExtension.java#L67)):*
> ```Java
>         // Remove current extension as active drag source from the UI
>         dragEndListenerHandle = addDragEndListener(
>                 event -> getUI().setActiveDragSource(null));
> ```

Could move these to their own protected method, `initListeners` that can be then overridden if necessary

---

*[server/src/main/java/com/vaadin/event/dnd/DragSourceExtension.java, line 81 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIrpGBuPNSZlqm45dk:-KgIrpGBuPNSZlqm45dl:bcmek33) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/event/dnd/DragSourceExtension.java#L81)):*
> ```Java
>             public void dragStart() {
>                 DragStartEvent<T> event = new DragStartEvent<>(target,
>                         getState(false).dataTransferText,
> ```

Why is _dataTransferText_ given for the drag start event ? Are we using it in some use case ? If not, just please remove ? the user can get it via the source component ?

---

*[server/src/main/java/com/vaadin/event/dnd/grid/GridDragEndEvent.java, line 27 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIsfi1DKStPKgdkqIz:-KgIsfi1DKStPKgdkqJ-:bcszq0n) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/event/dnd/grid/GridDragEndEvent.java#L27)):*
> Server side

Just please remove this, unnecessary

---

*[server/src/main/java/com/vaadin/event/dnd/grid/GridDragEndEvent.java, line 45 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIt5IhEHJWeMA2U1sR:-KgIt5IhEHJWeMA2U1sS:b8elpfi) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/event/dnd/grid/GridDragEndEvent.java#L45)):*
> dataTransferText
>      *         Data of type {@code "text"} from the {@code DataTransfer}
>      *         object.

Same as with `dragstart` event, is this now from the state, or from the client side drop event, if the first, we should probably just remove it.

---

*[server/src/main/java/com/vaadin/event/dnd/grid/GridDragEndListener.java, line 38 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIs96X6SC4GfAigkAP:-KgIs96X6SC4GfAigkAQ:b4zxjo) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/event/dnd/grid/GridDragEndListener.java#L38)):*
> ```Java
> 
>     /**
>      * Called when dragend is fired on a Grid row.
> ```

"Invoked when the user has dropped the dragged grid rows, or canceled the drag."

---

*[server/src/main/java/com/vaadin/event/dnd/grid/GridDragEndListener.java, line 41 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIsLQ81GLzcKzCZywZ:-KgIsLQ81GLzcKzCZyw_:boefrcb) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/event/dnd/grid/GridDragEndListener.java#L41)):*
> Server side

You should not have to mention "server side" or "client side" anywhere, unless specifically explaining difference in something between the server and client side.

---

*[server/src/main/java/com/vaadin/event/dnd/grid/GridDragStartEvent.java, line 27 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIqTjKkTUO8KuooISy:-KgIqTjKkTUO8KuooISz:buwjmov) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/event/dnd/grid/GridDragStartEvent.java#L27)):*
> ```Java
> 
> /**
>  * Server side drop event on an HTML5 drop target {@link Grid} row.
> ```

not drop

Also remove the server side, unnecessary 

---

*[server/src/main/java/com/vaadin/event/dnd/grid/GridDragStartEvent.java, line 43 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIq_oHiZqnH5i1Ayei:-KgIq_oHiZqnH5i1Ayej:b-y3nw4t) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/event/dnd/grid/GridDragStartEvent.java#L43)):*
> ```Java
>      *
>      * @param source
>      *         Grid component in which the items are dragged.
> ```

This sentence has some issues... how 'bout

"the source grid where the rows are being dragged from"

---

*[server/src/main/java/com/vaadin/event/dnd/grid/GridDragStartEvent.java, line 44 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIrZfJjwTrUxmG67M6:-KgIrZfJjwTrUxmG67M7:b3qxnvf) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/event/dnd/grid/GridDragStartEvent.java#L44)):*
> ```Java
>      * @param source
>      *         Grid component in which the items are dragged.
>      * @param dataTransferText
> ```

Can you remind, me why is dataTransferText given in _dragstart_ event here ??

---

*[server/src/main/java/com/vaadin/event/dnd/grid/GridDragStartListener.java, line 23 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIrNu7zaF78U3AnQGF:-KgIrNu7zaF78U3AnQGG:b-9qfzkv) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/event/dnd/grid/GridDragStartListener.java#L23)):*
> ```Java
> 
> /**
>  * Drop listener for HTML5 drop on a Grid row.
> ```

not drop

---

*[server/src/main/java/com/vaadin/event/dnd/grid/GridDragStartListener.java, line 38 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIrRKZkyNXS0Vu3U3_:-KgIrRKZkyNXS0Vu3U3a:b-158dc) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/event/dnd/grid/GridDragStartListener.java#L38)):*
> ```Java
> 
>     /**
>      * Called when dragstart is fired on a Grid row.
> ```

"Invoked when the user has started dragging grid's rows."

---

*[server/src/main/java/com/vaadin/ui/GridDragSourceExtension.java, line 82 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgItUCVmfzWNP3fsixG:-KgItUCVmfzWNP3fsixH:bk9ld7f) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/ui/GridDragSourceExtension.java#L82)):*
> ```Java
>                 GridDragStartEvent<T> event = new GridDragStartEvent<>(target,
>                         getState(false).dataTransferText,
>                         getState(false).effectAllowed, draggedItemKeys);
> ```

Why not just resolve the dragged item keys here, or better yet, in a specific method, and just pass the items to the event object, instead of the keys

Event objects should preferably have as less logic as possible

---

*[server/src/main/java/com/vaadin/ui/GridDragSourceExtension.java, line 138 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIts4005mzL3lDD_hC:-KgIts41KlOAeVLCagqp:b6vbhxi) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/ui/GridDragSourceExtension.java#L138)):*
> {@link
>      * GridDragStartListener#dragStart(GridDragStartEvent)} is called when
>      * dragstart event happens on the client side.

I think this is totally unnecessary. Could just add`@see GridDragStartEvent`

---

*[server/src/main/java/com/vaadin/ui/GridDragSourceExtension.java, line 154 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIu46a8m0zXDymxgN1:-KgIu46a8m0zXDymxgN2:b-xq4ykr) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/ui/GridDragSourceExtension.java#L154)):*
> {@link
>      * GridDragEndListener#dragEnd(GridDragEndEvent)} is called when
>      * dragend event happens on the client side.

replace with `@see GridDragEndEvent` 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8950)*
<!-- Sent from Reviewable.io -->

wbadam committed:
Fix review issues

wbadam commented:
Review status: 6 of 16 files reviewed at latest revision, 20 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIsoOFmE7J9IajLQ3v:-KgKC5DN2Ci1JLkdyOEs:b-dcib4r) (no related file):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

Please remove all mentions of "Server side something", those are IMO totally unnecessary, you should not have to make to distinctions unless you're explaining some difference in between server and client side.
</blockquote></details>

Changed all server side mentioned in javadocs

---

*[client/src/main/java/com/vaadin/client/connectors/grid/GridDragSourceExtensionConnector.java, line 102 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIkb0cSK8gN86qKke9:-KgJPmaBpKZtehiKY2zI:b3bug6x) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/client/src/main/java/com/vaadin/client/connectors/grid/GridDragSourceExtensionConnector.java#L102)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

Is this really necessary, or should the keys be cached after the drag has started?

Just thinking that what if the grid has been scrolled while dragging, thus the keys could have been removed from client side grid cache, at least in theory.
</blockquote></details>

Fair point

---

*[client/src/main/java/com/vaadin/client/connectors/grid/GridDragSourceExtensionConnector.java, line 109 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIkUiFOZbV870nRmfL:-KgJRCrLkVEwtYycS_JF:bae5b9j) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/client/src/main/java/com/vaadin/client/connectors/grid/GridDragSourceExtensionConnector.java#L109)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

> dragStartEvent

It seems to me that this is not always a drag start event now since it is used for drag ends too
</blockquote></details>

Removed the code from dragEnd that was using this method.

---

*[client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java, line 140 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgImfTLI_YdB3lHCJam:-KgJuclHYEM-PempBzt3:b-onme4) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java#L140)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

Can you isolate this to its own method that can be overridden by extending classes ?

Then you can move from `GridDragSourceExtensionConnector` this bit from `sendDragStartEventToServer`

```java 
        // Collect dragged rows and add their keys to the list
        getDraggedRows(dragStartEvent).forEach(
                row -> draggedItemKeys.add(row.get(GridState.JSONKEY_ROWKEY)));
```
</blockquote></details>

Not sure what exactly you mean, I created `createDataTransferText()` method, moved transfer data creation in both classes there and moved `draggedItemKeys` creation to `onDragStart()`

---

*[client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java, line 151 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgImDQz5hOFCglxpuhz:-KgJw5So9G7PRtc27KbY:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java#L151)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

You should probably document that this method is called ONLY IF there is a server side drag start event handler used
</blockquote></details>

Done.

---

*[client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java, line 181 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIn2jpPlEYiqMYOTEQ:-KgJwE2NfDJ_lmtXIjJD:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java#L181)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

> start

copy paste error
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/event/dnd/DragSourceExtension.java, line 67 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIogQX5BBDGONSZonq:-KgJwr4JxtY_SuBfpPE8:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/event/dnd/DragSourceExtension.java#L67)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

Could move these to their own protected method, `initListeners` that can be then overridden if necessary
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/event/dnd/DragSourceExtension.java, line 81 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIrpGBuPNSZlqm45dk:-KgJxeGz6q9fQN9n2MvG:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/event/dnd/DragSourceExtension.java#L81)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

Why is _dataTransferText_ given for the drag start event ? Are we using it in some use case ? If not, just please remove ? the user can get it via the source component ?
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/event/dnd/grid/GridDragEndEvent.java, line 27 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIsfi1DKStPKgdkqIz:-KgJy6O3y6VIe_awyYBe:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/event/dnd/grid/GridDragEndEvent.java#L27)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

> Server side

Just please remove this, unnecessary
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/event/dnd/grid/GridDragEndEvent.java, line 45 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIt5IhEHJWeMA2U1sR:-KgJz1ez4P0VN-m21_Rn:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/event/dnd/grid/GridDragEndEvent.java#L45)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

> dataTransferText
>      *         Data of type {@code "text"} from the {@code DataTransfer}
>      *         object.

Same as with `dragstart` event, is this now from the state, or from the client side drop event, if the first, we should probably just remove it.
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/event/dnd/grid/GridDragEndListener.java, line 38 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIs96X6SC4GfAigkAP:-KgJzHspsVUeZ_nz8Sap:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/event/dnd/grid/GridDragEndListener.java#L38)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

"Invoked when the user has dropped the dragged grid rows, or canceled the drag."
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/event/dnd/grid/GridDragEndListener.java, line 41 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIsLQ81GLzcKzCZywZ:-KgJzKUnQ_6sr-wrVxZf:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/event/dnd/grid/GridDragEndListener.java#L41)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

> Server side

You should not have to mention "server side" or "client side" anywhere, unless specifically explaining difference in something between the server and client side.
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/event/dnd/grid/GridDragStartEvent.java, line 27 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIqTjKkTUO8KuooISy:-KgJzVBvoyETqvPr0VeX:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/event/dnd/grid/GridDragStartEvent.java#L27)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

not drop

Also remove the server side, unnecessary 
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/event/dnd/grid/GridDragStartEvent.java, line 43 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIq_oHiZqnH5i1Ayei:-KgK-C7Qnxjbgy08obBL:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/event/dnd/grid/GridDragStartEvent.java#L43)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

This sentence has some issues... how 'bout

"the source grid where the rows are being dragged from"
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/event/dnd/grid/GridDragStartEvent.java, line 44 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIrZfJjwTrUxmG67M6:-KgK-HkQ2QKawpdDZjB3:ba3dfob) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/event/dnd/grid/GridDragStartEvent.java#L44)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

Can you remind, me why is dataTransferText given in _dragstart_ event here ??
</blockquote></details>

Removed

---

*[server/src/main/java/com/vaadin/event/dnd/grid/GridDragStartListener.java, line 23 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIrNu7zaF78U3AnQGF:-KgK-gA_60dH0pakEWfj:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/event/dnd/grid/GridDragStartListener.java#L23)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

not drop
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/event/dnd/grid/GridDragStartListener.java, line 38 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIrRKZkyNXS0Vu3U3_:-KgK-nI2qDVrJhI0PeMG:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/event/dnd/grid/GridDragStartListener.java#L38)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

"Invoked when the user has started dragging grid's rows."
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/ui/GridDragSourceExtension.java, line 82 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgItUCVmfzWNP3fsixG:-KgK2gm5gpwgCExwHMa0:biiqtwb) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/ui/GridDragSourceExtension.java#L82)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

Why not just resolve the dragged item keys here, or better yet, in a specific method, and just pass the items to the event object, instead of the keys

Event objects should preferably have as less logic as possible
</blockquote></details>

Done. Didn't know that it's better outside of the event.

---

*[server/src/main/java/com/vaadin/ui/GridDragSourceExtension.java, line 138 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIts4005mzL3lDD_hC:-KgK3AV6PfRGBFxdzOvr:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/ui/GridDragSourceExtension.java#L138)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

> {@link
>      * GridDragStartListener#dragStart(GridDragStartEvent)} is called when
>      * dragstart event happens on the client side.

I think this is totally unnecessary. Could just add`@see GridDragStartEvent`
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/ui/GridDragSourceExtension.java, line 154 at r1](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgIu46a8m0zXDymxgN1:-KgK3FYdsS6U27gaPL1z:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/5f31b4cc3445cf93d1f82c481fbec5263d5137ee/server/src/main/java/com/vaadin/ui/GridDragSourceExtension.java#L154)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

> {@link
>      * GridDragEndListener#dragEnd(GridDragEndEvent)} is called when
>      * dragend event happens on the client side.

replace with `@see GridDragEndEvent` 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8950)*
<!-- Sent from Reviewable.io -->

pleku commented:
Reviewed 10 of 10 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[client/src/main/java/com/vaadin/client/connectors/grid/GridDragSourceExtensionConnector.java, line 79 at r2](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgNqXuUfQrTbfJcmjiU:-KgNqXuUfQrTbfJcmjiV:b-qq1d3) ([raw file](https://github.com/vaadin/framework/blob/48b874aac7f835bb7f12621104a394a7714a3fe9/client/src/main/java/com/vaadin/client/connectors/grid/GridDragSourceExtensionConnector.java#L79)):*
> ```Java
>         draggedItemKeys = getDraggedRows(event).stream()
>                 .map(row -> row.getString(GridState.JSONKEY_ROWKEY))
>                 .collect(Collectors.toList());
> ```

Should there be some verification that there actually is dragged rows ?

If I have a cell that has a Label component in it that is also draggable, will the event propagate here and possibly cause that there are no rows ?

Could e.g. not do anything here if there are no dragged rows ? 

---

*[client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java, line 155 at r2](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgNpduElQF9rqBhiEBy:-KgNpduElQF9rqBhiEBz:b-bk2dhj) ([raw file](https://github.com/vaadin/framework/blob/48b874aac7f835bb7f12621104a394a7714a3fe9/client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java#L155)):*
> ```Java
>      * @param dragStartEvent
>      *         Event to set the data for.
>      * @return Textual data to be set for the event.
> ```

Can this be null if I don't want to set anything there ?

---

*[server/src/main/java/com/vaadin/ui/GridDragSourceExtension.java, line 105 at r2](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgO0FIvWOhWqInsRvQA:-KgO0FIvWOhWqInsRvQB:b-591ecl) ([raw file](https://github.com/vaadin/framework/blob/48b874aac7f835bb7f12621104a394a7714a3fe9/server/src/main/java/com/vaadin/ui/GridDragSourceExtension.java#L105)):*
> ```Java
>      * Collects the dragged items of a Grid given the list of item keys.
>      */
>     private Set<T> getDraggedItems(Grid<T> grid, List<String> draggedItemKeys) {
> ```

I think you should verify for null collection, as it is cleared on drag end in client side by setting it to null.

E.g.
```java
Objects.requireNonNull("Dragged item keys cannot be null", draggedItemKeys);
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8950)*
<!-- Sent from Reviewable.io -->

wbadam commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[client/src/main/java/com/vaadin/client/connectors/grid/GridDragSourceExtensionConnector.java, line 79 at r2](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgNqXuUfQrTbfJcmjiU:-KgOpHG4vHvoPgGdL-Nl:b4tx5sz) ([raw file](https://github.com/vaadin/framework/blob/48b874aac7f835bb7f12621104a394a7714a3fe9/client/src/main/java/com/vaadin/client/connectors/grid/GridDragSourceExtensionConnector.java#L79)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

Should there be some verification that there actually is dragged rows ?

If I have a cell that has a Label component in it that is also draggable, will the event propagate here and possibly cause that there are no rows ?

Could e.g. not do anything here if there are no dragged rows ? 
</blockquote></details>

We stop propagation on the target, perhaps should stop on the source as well?

Currently I think it doesn't cause troubles if the list of dragged rows is empty, but of course it can be verified with a simple condition here.

---

*[client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java, line 155 at r2](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgNpduElQF9rqBhiEBy:-KgOtWDNmU-NVcfUGW-L:b-kjikjf) ([raw file](https://github.com/vaadin/framework/blob/48b874aac7f835bb7f12621104a394a7714a3fe9/client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java#L155)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

Can this be null if I don't want to set anything there ?
</blockquote></details>

Yes, in that case the event will provide `null` for `getDataTransferText()`

---

*[server/src/main/java/com/vaadin/ui/GridDragSourceExtension.java, line 105 at r2](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgO0FIvWOhWqInsRvQA:-KgOuDrBtGM8zZj33n5d:b-aplv65) ([raw file](https://github.com/vaadin/framework/blob/48b874aac7f835bb7f12621104a394a7714a3fe9/server/src/main/java/com/vaadin/ui/GridDragSourceExtension.java#L105)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

I think you should verify for null collection, as it is cleared on drag end in client side by setting it to null.

E.g.
```java
Objects.requireNonNull("Dragged item keys cannot be null", draggedItemKeys);
```
</blockquote></details>

That's true. Is it better to verify it isn't null here or not sending null collection at all (but rather empty collection)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8950)*
<!-- Sent from Reviewable.io -->

pleku commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[client/src/main/java/com/vaadin/client/connectors/grid/GridDragSourceExtensionConnector.java, line 79 at r2](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgNqXuUfQrTbfJcmjiU:-KgTAmauFQXlNJQjMlBq:b-t23i3f) ([raw file](https://github.com/vaadin/framework/blob/48b874aac7f835bb7f12621104a394a7714a3fe9/client/src/main/java/com/vaadin/client/connectors/grid/GridDragSourceExtensionConnector.java#L79)):*
<details><summary><i>Previously, wbadam (Adam Wagner) wrote…</i></summary><blockquote>

We stop propagation on the target, perhaps should stop on the source as well?

Currently I think it doesn't cause troubles if the list of dragged rows is empty, but of course it can be verified with a simple condition here.
</blockquote></details>

As discussed, please stop propagation on drag start too

---

*[client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java, line 155 at r2](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgNpduElQF9rqBhiEBy:-KgTAiwiVK3o7OtX09IX:b-qftrbm) ([raw file](https://github.com/vaadin/framework/blob/48b874aac7f835bb7f12621104a394a7714a3fe9/client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java#L155)):*
<details><summary><i>Previously, wbadam (Adam Wagner) wrote…</i></summary><blockquote>

Yes, in that case the event will provide `null` for `getDataTransferText()`
</blockquote></details>

Update javadocs ?

---

*[server/src/main/java/com/vaadin/ui/GridDragSourceExtension.java, line 105 at r2](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgO0FIvWOhWqInsRvQA:-KgTAsjAeBb75C9jShQY:b-2bgafn) ([raw file](https://github.com/vaadin/framework/blob/48b874aac7f835bb7f12621104a394a7714a3fe9/server/src/main/java/com/vaadin/ui/GridDragSourceExtension.java#L105)):*
<details><summary><i>Previously, wbadam (Adam Wagner) wrote…</i></summary><blockquote>

That's true. Is it better to verify it isn't null here or not sending null collection at all (but rather empty collection)?
</blockquote></details>

Might be good to catch it here and throw an error, e.g. `IllegalStateException` if there are no dragged items received in the RPC call

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8950)*
<!-- Sent from Reviewable.io -->

wbadam committed:
Fix more review issues

wbadam commented:
Review status: 14 of 16 files reviewed at latest revision, 2 unresolved discussions.

---

*[client/src/main/java/com/vaadin/client/connectors/grid/GridDragSourceExtensionConnector.java, line 79 at r2](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgNqXuUfQrTbfJcmjiU:-KgTDvgU8YTIle8Vns-J:b-p49ccm) ([raw file](https://github.com/vaadin/framework/blob/48b874aac7f835bb7f12621104a394a7714a3fe9/client/src/main/java/com/vaadin/client/connectors/grid/GridDragSourceExtensionConnector.java#L79)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

As discussed, please stop propagation on drag start too
</blockquote></details>

Done on the generic level (`DragSourceExtensionConnector`)

---

*[client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java, line 155 at r2](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgNpduElQF9rqBhiEBy:-KgTEybDAZbrxvxJ8ngq:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/48b874aac7f835bb7f12621104a394a7714a3fe9/client/src/main/java/com/vaadin/client/extensions/DragSourceExtensionConnector.java#L155)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

Update javadocs ?
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/ui/GridDragSourceExtension.java, line 105 at r2](https://reviewable.io:443/reviews/vaadin/framework/8950#-KgO0FIvWOhWqInsRvQA:-KgTHXYbIzW0xfJ4kwzl:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/48b874aac7f835bb7f12621104a394a7714a3fe9/server/src/main/java/com/vaadin/ui/GridDragSourceExtension.java#L105)):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

Might be good to catch it here and throw an error, e.g. `IllegalStateException` if there are no dragged items received in the RPC call
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8950)*
<!-- Sent from Reviewable.io -->

pleku commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8950)*
<!-- Sent from Reviewable.io -->
